docs/internal/DistributedArchitectureGuide.md

```diff
@@ -365,18 +365,150 @@ There are several more Decider Services, implementing the `AutoscalingDeciderSer
+
+### Task tracking and registration
+
+Tasks are tracked in-memory on each node in the node's [TaskManager], new tasks are registered via one of the [TaskManager#register] methods.
+Registration of a task creates a [Task] instance with a unique-for-the-node numeric identifier, populates it with some metadata and stores it in the [TaskManager].
+
+The [register][TaskManager#register] methods will return the registered [Task] instance, which can be used to interact with the task. The [Task] class is often sub-classed to include task-specific data and operations. Specific [Task] subclasses are created by overriding the [createTask][TaskAwareRequest#createTask] method on the [TaskAwareRequest] passed to the [TaskManager#register] methods.
+
+When a task is completed, it must be unregistered via [TaskManager#unregister].
+
+#### A note about task IDs
+The IDs given to a task are numeric, supplied by a counter that starts at zero and increments over the life of the node process. So while they are unique in the individual node process, they would collide with IDs allocated after the node restarts, or IDs allocated on other nodes.
+
+To better identify a task in the cluster scope, the string `{node-ID}:{local-task-ID}` (e.g. `oTUltX4IQMOUUVeiohTt8A:124`) is used, however for the reasons explained this tuple will only be guaranteed unique for current tasks, not historical ones. [TaskId] is the DTO used to represent this tuple.
 
 ### What Tasks Are Tracked
 
-### Tracking A Task Across Threads
+The purpose of tasks is to provide management and visibility of cluster workload. There is some overhead involved in tracking a task, so they are best suited to tracking non-trivial and/or long-running operations. For smaller, more trivial operations visibility is probably better implemented using telemetry APIs.
+
+Some examples of operations that are tracked using tasks include:
+- Execution of [TransportAction]s
+  - [NodeClient#executeLocally] invokes [TaskManager#registerAndExecute]
+  - [RequestHandlerRegistry#processMessageReceived] registers tasks for actions that are spawned to handle [TransportRequest]s
+- Publication of cluster state updates
+
+### Tracking a Task Across Threads and Nodes
+
+#### ThreadContext
+
+[ThreadContext]:https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java
+[ThreadPool]:https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
+[ExecutorService]:https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ExecutorService.html
+
+All [ThreadPool] threads have an associated [ThreadContext]. The [ThreadContext] contains a map of headers which carry information relevant to the operation currently being executed. For example, a thread spawned to handle a REST request will include the HTTP headers received in that request.
```

Comment:
Might it be better to say ExecutorService here rather than ThreadPool?  Confusingly, ThreadPool is really just a collection of ExecutorServices, which each have the actual thread pools -- IIUC. The next paragraph better aligns with this.

Reply:
I think we need to be specific that we're discussing threads originating from `ExecutorService`s that are defined in `ThreadPool`, because as the `ThreadContext` docs mention:
https://github.com/elastic/elasticsearch/blob/a8dcce5db8a2f26ad1d83d4290c2a4b833a96f16/server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java#L47-L51
I don't know the details but it sounds like using `ThreadContext` with arbitrary threads might be more difficult/manual?

Reply:
Okay, it does seem like we use 'ThreadPool threads' elsewhere 👍